### ISSUE TITLE
Auto complete on analysis option not working correctly

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/sdk#59827]
### Working as intented

If you have on `analysis_options.yaml` the following:

```yaml
include: package:lints/recommended.yaml

linter:
  rules:
    - ^

analyzer:
  enable-experiment:
    - ^
```

On either `^` it will show you the possible rules for lints or possible experiments we have.

If you trigger auto-complete (CTRL + Space) on the left of the file, it will show you the sections; analyzer, linter, code-style, etc.

---
### *NOT* Working as intended

If you do that with one indentation inside `analyzer` it still shows the same as the previous paragraph.

If you do that on  `^` below:

```yaml
analyzer:
  language:
    ^
```

Here is the output:

![image](https://github.com/user-attachments/assets/75990e69-b364-4f94-9a1a-cc6e8a63212e)

This is a mix of the possible entries for the `analyzer` section and the entire file-valid sections and (at the end because this is alphabetically sorted) the valid options for the `analyzer.language` section (`strict-inference`, `strict-casts`, etc).

---

Also, on another note, we can't do with the `analyzer.language` section what we can with the `linter.rules` section of posting them on a list. We need to add each one with `: true` (again, the auto-complete here is totally lost when it would show `true`/`false` options it shows a lot more). 

---

Related to https://github.com/dart-lang/sdk/issues/57034.

**Comment by FMorschel** [comment]
FYI @srawlins @DanTup @bwilkerson

**Comment by DanTup** [comment]
Do you have a complete file sample that triggers this? When I try reproducing, I see this:

![image](https://github.com/user-attachments/assets/05385573-5a03-41b3-8ac7-2dac97d3e7c4)

**Comment by DanTup** [comment]
Part of the difference might be due to me having `"editor.wordBasedSuggestions": "off"` off. By default, VS Code will add suggestions for all the words it finds in the document (I think this is poor behaviour myself). However, even with that enabled, the case above works correctly (but all the words from the document do show up in some other places).

**Comment by FMorschel** [comment]
Not on the same OS, will test it out later today on that same project. But there is still this problem with the parts for the `analyzer` scope:

![image](https://github.com/user-attachments/assets/2472ad64-f12a-4303-ba3c-36b56cfbbf1a)

---

Also,`wordBasedSuggestions` is set to `off` on Dart -> User

![image](https://github.com/user-attachments/assets/8d5e03ea-2259-4a73-af69-293a9bb3313d)

But I'm unsure of what it means. Would it only be set to `off` on `.dart` files?

**Comment by DanTup** [comment]
> But I'm unsure of what it means. Would it only be set to off on .dart files?

Correct. I have it set to `off` in my VS Code User Settings so it's disabled for all languages, because IMO it's never useful. I think if you turned it off globally (or for YAML), you might be seeing empty results in the places where you see those (I think it might only appear as a fallback if the server doesn't return anything?).

I can reproduce this one:

![image](https://github.com/user-attachments/assets/d982d197-0961-4092-99d1-7918c452dc68)

**Comment by FMorschel** [comment]
> Part of the difference might be due to me having `"editor.wordBasedSuggestions": "off"`

I`ve tested it again and I can confirm that was it.

I've also confirmed that the case you managed to repro above only happens if you already have one of the internal options for `analyzer` _later_. 

![image](https://github.com/user-attachments/assets/2bd0e5d1-4db0-4cb3-9272-5cd2e7e6e736)

If you trigger autocomplete after, it still thinks you're at the last spot:

![image](https://github.com/user-attachments/assets/50c7ce31-45cf-4fe8-b09b-8f902733eb13)

See that there is no auto-complete option for `true`/`false` here and as I mentioned at the start, for some reason, this is not coherent with the `linter->rules` section where we can use a yaml list.

![image](https://github.com/user-attachments/assets/a09c6ee3-ef36-4d74-b322-b4ddc8845916)

![image](https://github.com/user-attachments/assets/6f809585-258f-40fd-9351-e22d4f25542d)

**Comment by DanTup** [comment]
I had a quick look at this thinking it'd be a trivial fix, but I don't think it is. The reason it fails is that in code like:

```yaml
analyzer:
  ^
```

The AST for the file looks like `{analyzer: null}` and we return the `null` node which allows us to provide completion for the 1analyzer` section.

However, in the caseof:

```yaml
analyzer:
  ^
  foo: bar
```

Here, the AST looks like `{ analyzer: { foo: bar } }` and therefore there is no equivalent of the `null` node above, so we don't have a good node to base the completion on (we can't return `analyzer`, because completion for `analyzer` would still be the parent completions, which _include_ `analyzer` but are not children of it).

So I think it'll need a little reworking of `YamlCompletionResults._producerForPath` so that we're not just trying to compute completion based on some node, but support producing completion for the _children_ of a node (which probably also means representing this in some way in the `path`.. perhaps as a synthetic key?).

In the meantime, I did open a fix for the missing booleans for the `strict-` language options at https://dart-review.googlesource.com/c/sdk/+/409140 though.

**Comment by FMorschel** [comment]
Doing so would probably help with https://github.com/dart-lang/sdk/issues/57034, too.

Is your change enough for me to use the `strict-` language options as a list? Or is that something the team would like to avoid for some meaning @bwilkerson? I feel like these are close enough to how lints work today so they could be added as a list that would mean `true` for all.

**Comment by bwilkerson** [comment]
I believe that the medium to long term goal is to replace the language options with lints. I don't know what the time frame is for that, so I don't know whether it's worth the effort.

@srawlins

**Comment by DanTup** [comment]
> Is your change enough for me to use the strict- language options as a list?

No, all I did was make the completion for the existing (map) entries provide boolean completions instead of no completions.

Since it just occurred to me, I'll also note the original issue described above also affects Pubspec (eg. you get "dependencies:" in completion here instead of package names):

![Image](https://github.com/user-attachments/assets/b83c7e50-0211-4493-aed4-2cd82110c6e1)

**Comment by FMorschel** [comment]
I'm guessing that https://github.com/dart-lang/sdk/issues/57077 is a duplicate I didn't find originally. Can someone else confirm?

**Comment by DanTup** [comment]
Yep, looks like the same issue to me.

**Comment by FMorschel** [comment]
I've tried fixing https://github.com/dart-lang/sdk/issues/61544, and the CL also fixes this.

https://dart-review.googlesource.com/c/sdk/+/511260

<img width="585" height="466" alt="Image" src="https://github.com/user-attachments/assets/2eee0e1b-01e2-46b8-a67f-cf1ac1ba60ee" />

But I still see:

<img width="522" height="562" alt="Image" src="https://github.com/user-attachments/assets/5d5e5239-de6f-4c95-ab47-96fa83b487fa" />

This seems like:

<img width="886" height="608" alt="Image" src="https://github.com/user-attachments/assets/8ac8c4e9-8930-4fef-ac6e-c24de07564eb" />

I have a fix for this too, which considers the indent for the completions. I'll upload it once the other one lands, since some of the tests and implementation are based on the assumption that the fixes the above-mentioned CL did are working.

I'll be sure to close this and https://github.com/dart-lang/sdk/issues/57077 on that new CL.

FYI @bwilkerson @srawlins

**Comment by FMorschel** [comment]
Here is the new CL https://dart-review.googlesource.com/c/sdk/+/516000
